### ISSUE TITLE
Why object is not detected on second detection in jetpack5 docker container

### ISSUE THREAD

**Issue by Chigo55** [ultralytics/ultralytics#21015]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Using the ultralytics/ultralytics:latest-jetson-jetpack5 docker image on the Jetson AGX Xavier board, I created a container using the following command and copied the test code, model, and image to the container to execute the test code

Create docker containers and copy files clearly
docker run -dt --name test --runtime nvidia --gpus all ultralytics:jetpack5
docker cp /home/xaviernx/xxx/. test:/ultralytics/
docker exec -it test /bin/bash

Test Code
```
import torch
from ultralytics import YOLO

model = YOLO(model="checkpoints/yolo_augment.pt")

device = "cuda:0" if torch.cuda.is_available() else "cpu"
print(device)

results = model.predict(
    source="video/feed_summery.mp4",
    verbose=True,
    stream=True,
    device=device,
    conf=0.1,
)

for result in results:
    pass
```

The test code initially confirmed that the object was detected normally, and the second attempt at detection with the same code did not detect any object. Even though the same model, same image, and same code were used. Later, I found that only the log was not detected, and I modified the code to store the detected object to confirm that the object was detected normally. 

Modified test code 
```
import torch
from ultralytics import YOLO

torch.cuda.empty_cache()

model = YOLO(model="checkpoints/yolo_augment.pt")

device = "cuda:0" if torch.cuda.is_available() else "cpu"
print(device)

results = model.predict(
    source="video/feed_summery.mp4",
    verbose=True,
    stream=True,
    save=True,
    save_txt=True,
    device=device,
    conf=0.1,
)

for result in results:
    pass
```

No objects were detected when the modified test code was executed, and no objects were saved as text files.

Can you help me with these issues?

### Environment

Ultralytics 8.3.150 🚀 Python-3.8.10 torch-2.2.0 CUDA:0 (Xavier, 6833MiB)
Setup complete ✅ (6 CPUs, 6.7 GB RAM, 35.6/57.7 GB disk)

OS                  Linux-5.10.216-tegra-aarch64-with-glibc2.29
Environment         Docker
Python              3.8.10
Install             git
Path                /ultralytics/ultralytics
RAM                 6.67 GB
Disk                35.6/57.7 GB
CPU                 ARMv8 Processor rev 0 (v8l)
CPU count           6
GPU                 Xavier, 6833MiB
GPU count           1
CUDA                11.4

numpy               ✅ 1.24.4>=1.23.0
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.2.0>=1.8.0
torch               ✅ 2.2.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.17.2+c1d70fe>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.0.3>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0
{'OS': 'Linux-5.10.216-tegra-aarch64-with-glibc2.29', 'Environment': 'Docker', 'Python': '3.8.10', 'Install': 'git', 'Path': '/ultralytics/ultralytics', 'RAM': '6.67 GB', 'Disk': '35.6/57.7 GB', 'CPU': 'ARMv8 Processor rev 0 (v8l)', 'CPU count': 6, 'GPU': 'Xavier, 6833MiB', 'GPU count': 1, 'CUDA': '11.4', 'Package Info': {'numpy': '✅ 1.24.4>=1.23.0', 'matplotlib': '✅ 3.7.5>=3.3.0', 'opencv-python': '✅ 4.11.0.86>=4.6.0', 'pillow': '✅ 10.4.0>=7.1.2', 'pyyaml': '✅ 6.0.2>=5.3.1', 'requests': '✅ 2.32.3>=2.23.0', 'scipy': '✅ 1.10.1>=1.4.1', 'torch': '✅ 2.2.0!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.17.2+c1d70fe>=0.9.0', 'tqdm': '✅ 4.67.1>=4.64.0', 'psutil': '✅ 7.0.0', 'py-cpuinfo': '✅ 9.0.0', 'pandas': '✅ 2.0.3>=1.1.4', 'ultralytics-thop': '✅ 2.0.14>=2.0.0'}}

### Minimal Reproducible Example

video 1/1 (frame 2469/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 25.3ms
video 1/1 (frame 2470/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 24.0ms
video 1/1 (frame 2471/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 24.1ms
video 1/1 (frame 2472/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 23.8ms
video 1/1 (frame 2473/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 25.2ms
video 1/1 (frame 2474/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 24.4ms
video 1/1 (frame 2475/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 24.4ms
video 1/1 (frame 2476/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 24.5ms
video 1/1 (frame 2477/10448) /ultralytics/video/feed_summery.mp4: 384x640 (no detections), 23.9ms

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Chigo55, thank you for reporting this and for providing such a thorough description and environment details! 🤖 This is an automated response to help you get started, and an Ultralytics engineer will also review your issue soon.

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for guidance, as many common questions are addressed there. For usage examples, see the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) sections.

Since this is a 🐛 Bug Report and you’ve included a minimal reproducible example (MRE), thank you! If possible, please confirm that your issue persists after upgrading to the latest version of the `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

For reference, YOLO can be run in several up-to-date verified environments:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Join our community for real-time chat and support on [Discord](https://discord.com/invite/ultralytics) 🎧, longer discussions on [Discourse](https://community.ultralytics.com/), or knowledge sharing on our [Subreddit](https://reddit.com/r/Ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your detailed report and patience while our team reviews your issue! 🛠️

**Comment by RizwanMunawar** [comment]
@lakshanthad Hi, can you please check this? Thanks

**Comment by lakshanthad** [comment]
@Chigo55 I will debug this and get back to you.

**Comment by lakshanthad** [comment]
After some debugging, It seems that the issue is present not only in Docker, but also in host system when running on JetPack 5. However, the issue is not there on JetPack 6 systems. 

I was able to go back in versions and isolate the issue. Seems the issue is originated from [this commit](https://github.com/ultralytics/ultralytics/commit/9605f31a2f783639d6d0c7be54fc3414834c0a43) in `v8.3.130`.
Detections are working as normal for JetPack 5 when I revert this commit on the latest main branch.

It could be the case where since JetPack 5 is running on CUDA 11.4, it does not support this optimized model fusion operation, whereas JetPack 6 is running on CUDA 12.1. This could be debugged further by installing CUDA 11.x on a X86 GPU system to see whether the issue persists.

Any thoughts on this? @Y-T-G @Laughing-q  

On `v8.3.129` without optimized model fusion operation

```sh
(env) nvidia@ubuntu:~/YOLO11/ultralytics$ yolo detect predict
WARNING ⚠️ 'source' argument is missing. Using default 'source=/home/nvidia/YOLO11/ultralytics/ultralytics/assets'.
Ultralytics 8.3.129 🚀 Python-3.8.10 torch-2.2.0 CUDA:0 (Orin, 62800MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

image 1/2 /home/nvidia/YOLO11/ultralytics/ultralytics/assets/bus.jpg: 640x480 4 persons, 1 bus, 204.6ms
image 2/2 /home/nvidia/YOLO11/ultralytics/ultralytics/assets/zidane.jpg: 384x640 2 persons, 1 tie, 204.5ms
Speed: 4.0ms preprocess, 204.6ms inference, 3.6ms postprocess per image at shape (1, 3, 384, 640)
```

On `v8.3.130` with optimized model fusion operation

```sh
(env) nvidia@ubuntu:~/YOLO11/ultralytics$ yolo detect predict
WARNING ⚠️ 'source' argument is missing. Using default 'source=/home/nvidia/YOLO11/ultralytics/ultralytics/assets'.
Ultralytics 8.3.130 🚀 Python-3.8.10 torch-2.2.0 CUDA:0 (Orin, 62800MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

image 1/2 /home/nvidia/YOLO11/ultralytics/ultralytics/assets/bus.jpg: 640x480 (no detections), 206.6ms
image 2/2 /home/nvidia/YOLO11/ultralytics/ultralytics/assets/zidane.jpg: 384x640 (no detections), 205.9ms
Speed: 4.0ms preprocess, 206.2ms inference, 2.0ms postprocess per image at shape (1, 3, 384, 640)
```

**Comment by Y-T-G** [comment]
@lakshanthad Does the issue occur with CPU prediction on Jetpack 5?

**Comment by lakshanthad** [comment]
@Y-T-G CPU prediction is all good on JetPack 5:

```sh
(env) nvidia@ubuntu:~/YOLO11-cpu/ultralytics$ yolo detect predict
WARNING ⚠️ 'source' argument is missing. Using default 'source=/home/nvidia/YOLO11-cpu/ultralytics/ultralytics/assets'.
Ultralytics 8.3.153 🚀 Python-3.8.10 torch-2.4.1 CPU (ARMv8 Processor rev 1 (v8l))
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

image 1/2 /home/nvidia/YOLO11-cpu/ultralytics/ultralytics/assets/bus.jpg: 640x480 4 persons, 1 bus, 297.3ms
image 2/2 /home/nvidia/YOLO11-cpu/ultralytics/ultralytics/assets/zidane.jpg: 384x640 2 persons, 1 tie, 244.7ms
Speed: 3.6ms preprocess, 271.0ms inference, 3.3ms postprocess per image at shape (1, 3, 384, 640)
```

**Comment by lakshanthad** [comment]
@Y-T-G On my RTX 3060 GPU after using CUDA11.x Docker container (because host system has CUDA 12.1)

```sh
root@078108a623b9:/ultralytics# yolo detect predict
Ultralytics 8.3.153 🚀 Python-3.8.10 torch-1.12.1+cu113 CUDA:0 (NVIDIA GeForce RTX 3060, 12011MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

image 1/2 /ultralytics/ultralytics/assets/bus.jpg: 640x480 4 persons, 1 bus, 8.9ms
image 2/2 /ultralytics/ultralytics/assets/zidane.jpg: 384x640 2 persons, 1 tie, 8.9ms
Speed: 0.9ms preprocess, 8.9ms inference, 0.8ms postprocess per image at shape (1, 3, 384, 640)
```

So it seems that the issue is not with CUDA version in general. Maybe it is with CUDA11.x specifically on ARM systems.

**Comment by Y-T-G** [comment]
@lakshanthad The commit only changes on which device the fusing occurs. So it seems like fusing on CPU and then moving to GPU doesn't work on JP5. What if you move the model to the GPU before the first prediction? Maybe that fixes it.

```
model.cuda()
results = model()
```

**Comment by lakshanthad** [comment]
@Y-T-G That seems to work!

**Comment by lakshanthad** [comment]
I have pushed a PR for this @Y-T-G. The logic works for JetPack 5 and below systems.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #21028, authored by lakshanthad with valuable contributions from Laughing-q and glenn-jocher.

Key updates from this PR include:
- Enhanced test reliability by adding checks to ensure model results are not empty.
- Improved model loading for PyTorch models by moving the model to the device before fusing, which maintains compatibility with NVIDIA JetPack 5 and earlier and can help avoid related errors.

These changes aim to make model deployment smoother and improve robustness, especially when working with custom or in-memory PyTorch models.

To test if the update resolves your issue, try installing the latest code directly from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, you can wait for the next official PyPI release.

Please let us know if this resolves your issue or if you encounter any further problems. Thank you for reporting this, and any additional feedback is always appreciated!
